pkg/api/v1/types.go

```diff
@@ -1217,12 +1217,14 @@ type EndpointSubset struct {
 	Ports     []EndpointPort    `json:"ports,omitempty" description:"port numbers available on the related IP addresses"`
 }
 
-// EndpointAddress is a tuple that describes single IP address.
+// EndpointAddress is a tuple that describes single endpoint address.
 type EndpointAddress struct {
-	// The IP of this endpoint.
-	// TODO: This should allow hostname or IP, see #4447.
+	// The IP of this endpoint, mutually exclusive with Hostname
 	IP string `json:"ip" description:"IP address of the endpoint"`
 
+	// Hostname for this endpoint, you can specify IP or Hostname, but not both.
+	Hostname string `json:"hostname" description:"Hostname for this endpoint"`
```

Comment:
Regarding one field vs two, this is a question we need to solve in a bunch of places.  We were originally going to make IP fields use a net.IP-derived type for strong validation, but we backed off when we realized some places really do want host or IP.  So in my mind this is still an open question.  This sets precedent - we either use twin fields all over or we document these all as being "ip or hostname" and stick with 'string'.  I have not put much thought into this, but I guess it doesn't matter much either way.

@bgrant0607 for API call